rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java

```diff
@@ -0,0 +1,315 @@
+                    Object next = it.next();
+                    if (validContexts.contains(next)) {
+                        replacements.put(typeCast, instanceOf);
+                        break;
+                    } else if (root == next) {
+                        break;
+                    }
+                }
+            }
+        }
+
+        public boolean isEmpty() {
+            return replacements.isEmpty();
+        }
+
+        public J.InstanceOf processInstanceOf(J.InstanceOf instanceOf, Cursor cursor) {
+            @Nullable JavaType type = ((TypeTree) instanceOf.getClazz()).getType();
+            String name = patternVariableName(instanceOf, cursor);
+            J.InstanceOf result = instanceOf.withPattern(new J.Identifier(
+                    randomId(),
+                    Space.build(" ", emptyList()),
+                    Markers.EMPTY,
+                    name,
+                    type,
+                    null));
+
+            // update entry in replacements to share the pattern variable name
+            for (Map.Entry<J.TypeCast, J.InstanceOf> entry : replacements.entrySet()) {
+                if (entry.getValue() == instanceOf) {
+                    entry.setValue(result);
+                }
+            }
+            return result;
+        }
+
+        @Nullable
+        public J processTypeCast(J.TypeCast typeCast, Cursor cursor) {
+            J.InstanceOf instanceOf = replacements.get(typeCast);
+            if (instanceOf != null && instanceOf.getPattern() != null) {
+                String name = ((J.Identifier) instanceOf.getPattern()).getSimpleName();
+                TypedTree owner = cursor.firstEnclosing(J.MethodDeclaration.class);
+                owner = owner != null ? owner : cursor.firstEnclosingOrThrow(J.ClassDeclaration.class);
+                JavaType.Variable fieldType = new JavaType.Variable(null, Flag.Default.getBitMask(), name, owner.getType(), typeCast.getType(), emptyList());
+                return new J.Identifier(
+                        randomId(),
+                        typeCast.getPrefix(),
+                        Markers.EMPTY,
+                        name,
+                        typeCast.getType(),
+                        fieldType);
+            }
+            return null;
+        }
+    }
+
+    private static String patternVariableName(J.InstanceOf instanceOf, Cursor cursor) {
+        return VariableNameUtils.generateVariableName(shortVariableBaseName((TypeTree) instanceOf.getClazz()), cursor, INCREMENT_NUMBER);
+    }
+
+    private static String shortVariableBaseName(TypeTree typeTree) {
+        String baseName = shortVariableBaseName(typeTree.getType());
+        return typeTree instanceof J.ArrayType ? baseName + 's' : baseName;
+    }
+
+    private static String shortVariableBaseName(@Nullable JavaType type) {
+        // the instanceof operator only accepts classes (without generics) and arrays
```

Comment:
Doesn't look like this handles arrays of arrays. e.g.: `String[][]`. Not sure how common it is that we would run into that.

Reply:
Good point. Let me add a test for this and see if it needs fixing.

Reply:
For a multi-dimensional array like `int[][]` the LST node is an `J.ArrayType` with `elementType = int` and `dimension` a list of size `2`. It appears to work correctly. Added a test case for this.